File: client/src/api/endpoints/apiKeys.ts
import { API_BASE_URL, getHeaders, handleApiError } from "../config";

export interface ApiKey {
  id: string;
  name: string;
  organization_id: string;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
  last_used_at?: string;
  usage_count: number;
  rate_limit_per_minute: number;
  webhook_url?: string;
  webhook_secret?: string;
  webhook_timeout: number;
  webhook_headers?: Record<string, string>;
  secret_key_preview: string;
  created_by_name: string;
}

export interface ApiKeyCreate {
  name: string;
  rate_limit_per_minute?: number;
  webhook_url?: string;
  webhook_secret?: string;
  webhook_timeout?: number;
  webhook_headers?: Record<string, string>;
}

export interface ApiKeySecret {
  id: string;
  name: string;
  secret_key: string;
}

export interface ApiKeyUpdate {
  name?: string;
  is_active?: boolean;
  rate_limit_per_minute?: number;
  webhook_url?: string;
  webhook_secret?: string;
  webhook_timeout?: number;
  webhook_headers?: Record<string, string>;
}

// Create a new API key
export const createApiKey = async (data: ApiKeyCreate): Promise<ApiKeySecret> => {
  const response = await fetch(`${API_BASE_URL}/api-keys/`, {
    method: "POST",
    headers: getHeaders(),
    body: JSON.stringify(data),
  });

  await handleApiError(response);
  return response.json();
};

// List all API keys
export const listApiKeys = async (): Promise<ApiKey[]> => {
  const response = await fetch(`${API_BASE_URL}/api-keys/`, {
    method: "GET",
    headers: getHeaders(),
  });

  await handleApiError(response);
  return response.json();
};

// Get a specific API key
export const getApiKey = async (id: string): Promise<ApiKey> => {
  const response = await fetch(`${API_BASE_URL}/api-keys/${id}`, {
    method: "GET",
    headers: getHeaders(),
  });

  await handleApiError(response);
  return response.json();
};

// Update an API key
export const updateApiKey = async (id: string, data: ApiKeyUpdate): Promise<ApiKey> => {
  const response = await fetch(`${API_BASE_URL}/api-keys/${id}`, {
    method: "PATCH",
    headers: getHeaders(),
    body: JSON.stringify(data),
  });

  await handleApiError(response);
  return response.json();
};

// Delete an API key
export const deleteApiKey = async (id: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/api-keys/${id}`, {
    method: "DELETE",
    headers: getHeaders(),
  });

  await handleApiError(response);
};

File: server/app/api/v1/endpoints/api_keys.py
"""
API Key management endpoints.
"""

from typing import List
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_current_organization, get_current_user, get_tenant_db
from app.models import Organization, User
from app.schemas import ApiKeyCreate, ApiKeyUpdate, ApiKeyResponse, ApiKeySecret
from app.services.api_key_service import ApiKeyService

router = APIRouter()


@router.post("/", response_model=ApiKeySecret, status_code=status.HTTP_201_CREATED)
async def create_api_key(
    api_key_data: ApiKeyCreate,
    organization: Organization = Depends(get_current_organization),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_tenant_db),
):
    """
    Create a new API key.
    
    Args:
        api_key_data: API key creation data
        organization: Current organization
        current_user: Current authenticated user
        db: Database session
        
    Returns:
        ApiKeySecret: Created API key with secret (shown only once)
    """
    return await ApiKeyService.create_api_key(
        api_key_data=api_key_data,
        organization_id=organization.id,
        created_by_id=current_user.id,
        db=db
    )


@router.get("/", response_model=List[ApiKeyResponse])
async def list_api_keys(
    skip: int = 0,
    limit: int = 100,
    organization: Organization = Depends(get_current_organization),
    db: AsyncSession = Depends(get_tenant_db),
):
    """
    List API keys for the organization.
    
    Args:
        skip: Number of records to skip
        limit: Maximum number of records to return
        organization: Current organization
        db: Database session
        
    Returns:
        List[ApiKeyResponse]: List of API keys
    """
    return await ApiKeyService.list_api_keys(
        organization_id=organization.id,
        db=db,
        skip=skip,
        limit=limit
    )


@router.get("/{api_key_id}", response_model=ApiKeyResponse)
async def get_api_key(
    api_key_id: UUID,
    organization: Organization = Depends(get_current_organization),
    db: AsyncSession = Depends(get_tenant_db),
):
    """
    Get a specific API key.
    
    Args:
        api_key_id: API key ID
        organization: Current organization
        db: Database session
        
    Returns:
        ApiKeyResponse: API key details
    """
    api_key = await ApiKeyService.get_api_key(
        api_key_id=api_key_id,
        organization_id=organization.id,
        db=db
    )
    
    if not api_key:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API key not found"
        )
    
    return api_key


@router.patch("/{api_key_id}", response_model=ApiKeyResponse)
async def update_api_key(
    api_key_id: UUID,
    api_key_data: ApiKeyUpdate,
    organization: Organization = Depends(get_current_organization),
    db: AsyncSession = Depends(get_tenant_db),
):
    """
    Update an API key.
    
    Args:
        api_key_id: API key ID
        api_key_data: API key update data
        organization: Current organization
        db: Database session
        
    Returns:
        ApiKeyResponse: Updated API key
    """
    api_key = await ApiKeyService.update_api_key(
        api_key_id=api_key_id,
        api_key_data=api_key_data,
        organization_id=organization.id,
        db=db
    )
    
    if not api_key:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API key not found"
        )
    
    return api_key


@router.delete("/{api_key_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_api_key(
    api_key_id: UUID,
    organization: Organization = Depends(get_current_organization),
    db: AsyncSession = Depends(get_tenant_db),
):
    """
    Delete an API key.
    
    Args:
        api_key_id: API key ID
        organization: Current organization
        db: Database session
    """
    success = await ApiKeyService.delete_api_key(
        api_key_id=api_key_id,
        organization_id=organization.id,
        db=db
    )
    
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="API key not found"
        )